src/Evolution/DiscontinuousGalerkin/Messages/BoundaryMessage.cpp

```diff
@@ -0,0 +1,216 @@
+
+  if (subcell_size != 0) {
+    // double* + 1 == char* + 8 because double* is 8 bytes
+    // Place subcell data right after dg pointer
+    out_msg->subcell_ghost_data =
+        reinterpret_cast<double*>(std::addressof(out_msg->dg_flux_data)) + 1;
+    memcpy(out_msg->subcell_ghost_data, in_msg->subcell_ghost_data,
+           subcell_size * sizeof(double));
+  }
+  if (dg_size != 0) {
+    // Place dg data right after subcell data
+    out_msg->dg_flux_data =
+        reinterpret_cast<double*>(std::addressof(out_msg->dg_flux_data)) + 1 +
+        subcell_size;
+    memcpy(out_msg->dg_flux_data, in_msg->dg_flux_data,
+           dg_size * sizeof(double));
+  }
+
+  // Gotta clean up
+  delete in_msg;
+  return static_cast<void*>(out_msg);
+}
+
+template <size_t Dim>
+BoundaryMessage<Dim>* BoundaryMessage<Dim>::unpack(void* in_buf) {
+  // We expect in_buf to be in a format where we can interpret it as a
+  // BoundaryMessage which is why we can immediately unpack the data as if it
+  // was a BoundaryMessage. All we have to do is set the pointers to the
+  // beginning of their respective data
+  BoundaryMessage<Dim>* buffer =
+      reinterpret_cast<BoundaryMessage<Dim>*>(in_buf);
+
+  const size_t subcell_size = buffer->subcell_ghost_data_size;
+  const size_t dg_size = buffer->dg_flux_data_size;
+
+  if (subcell_size != 0) {
+    // double* + 1 == char* + 8 because double* is 8 bytes
+    // Subcell data is located right after dg pointer
+    buffer->subcell_ghost_data =
+        reinterpret_cast<double*>(std::addressof(buffer->dg_flux_data)) + 1;
+  } else {
+    buffer->subcell_ghost_data = nullptr;
+  }
+  if (dg_size != 0) {
+    // Dg data is located right after subcell data
+    buffer->dg_flux_data =
+        reinterpret_cast<double*>(std::addressof(buffer->dg_flux_data)) + 1 +
+        subcell_size;
+  } else {
+    buffer->dg_flux_data = nullptr;
+  }
+
+  // We don't delete in_buf here because it is actually the data we want. We
+  // didn't do any new allocations/memcpy's so no need to clean up
+  return buffer;
+}
+
+template <size_t Dim>
+bool operator==(const BoundaryMessage<Dim>& lhs,
+                const BoundaryMessage<Dim>& rhs) {
+  const bool equal =
+      lhs.subcell_ghost_data_size == rhs.subcell_ghost_data_size and
+      lhs.dg_flux_data_size == rhs.dg_flux_data_size and
+      lhs.sent_across_nodes == rhs.sent_across_nodes and
+      lhs.current_time_step_id == rhs.current_time_step_id and
+      lhs.next_time_step_id == rhs.next_time_step_id and
+      lhs.volume_or_ghost_mesh == rhs.volume_or_ghost_mesh and
+      lhs.interface_mesh == rhs.interface_mesh;
+
+  // We check here first to avoid looping over the arrays if it's unnecessary.
+  if (not equal) {
+    return false;
+  }
+
+  // We are guaranteed that lhs.subcell_size == rhs.subcell_size and
+  // lhs.dg_size == rhs.dg_size at this point so it's safe to loop over
+  // everything
+  for (size_t i = 0; i < lhs.subcell_ghost_data_size; i++) {
```

Comment:
replace with `std::equal` and then you could stick it all into one line (up to you about the 1-liner).